docs: portuguese translation (#853)
t3-oss/create-t3-app: www/src/pages/pt/deployment/docker.md

---
title: Docker
description: Deploy com Docker
layout: ../../../layouts/docs.astro
lang: pt
---

Você pode colocar essa stack em contêineres e fazer deploy dela como um único contêiner usando o Docker ou como parte de um grupo de contêineres usando o docker-compose. Veja [`ajcwebdev/ct3a-docker`](https://github.com/ajcwebdev/ct3a-docker) para um repositório de exemplo baseado neste documento.

## Configuração do Projeto Docker

> Por favor, note que o Next.js requer um processo diferente para tempo de build (disponível no frontend, prefixado por `NEXT_PUBLIC`) e ambiente de tempo de execução, somente do lado do servidor, variáveis. Nesta demonstração estamos usando duas variáveis, preste atenção em suas posições no `Dockerfile`, argumentos de linha de comando e `docker-compose.yml`:

Reviewer: ```suggestion
Por favor, note que o Next.js requer um processo diferente para tempo de build (disponível no frontend, prefixado por `NEXT_PUBLIC`) e ambiente de tempo de execução, somente do lado do servidor, variáveis. Nesta demonstração estamos usando duas variáveis. Preste atenção em suas posições no `Dockerfile`, argumentos de linha de comando e `docker-compose.yml`:
```

- `DATABASE_URL` (usado pelo servidor)
- `NEXT_PUBLIC_CLIENTVAR` (usado pelo cliente)

> ### 1. Próxima configuração

Reviewer: ```suggestion
### 1. Configuração do Next
```

> Em seu arquivo [`next.config.mjs`](https://github.com/t3-oss/create-t3-app/blob/main/cli/template/base/next.config.mjs), configure a opção `output` como `standalone` para [reduzir o tamanho de imagens aproveitando automaticamente os rastreamentos de saída](https://nextjs.org/docs/advanced-features/output-file-tracing):

Reviewer 2: ```suggestion
Em seu arquivo [`next.config.mjs`](https://github.com/t3-oss/create-t3-app/blob/main/cli/template/base/next.config.mjs), configure a opção `output` como `standalone` para [reduzir o tamanho de imagens](https://nextjs.org/docs/advanced-features/output-file-tracing):
```

Over-information is not needed in translations, the target link already explains the process.

> Em seu [`next.config.mjs`](https://github.com/t3-oss/create-t3-app/blob/main/cli/template/base/next.config.mjs), adicione a ` configuração de opção de saída autônoma para [reduzir o tamanho da imagem aproveitando automaticamente os rastreamentos de saída](https://nextjs.org/docs/advanced-features/output-file-tracing):

Reviewer: ```suggestion
Em seu arquivo [`next.config.mjs`](https://github.com/t3-oss/create-t3-app/blob/main/cli/template/base/next.config.mjs), configure a opção `output` como `standalone` para [reduzir o tamanho de imagens aproveitando automaticamente os rastreamentos de saída](https://nextjs.org/docs/advanced-features/output-file-tracing):
```

I am not completely satisfied with the expression "aproveitando automaticamente os rastreamentos de saída". 
I do not think the current expression is clear enough in Portuguese to understand the meaning.
If someone has a better expression, please suggest.

Author: Liked that one, gonna commit

```diff
export default defineNextConfig({
  reactStrictMode: true,
  swcMinify: true,
+ output: "standalone",
});
```

> ### 2. Criar arquivo dockerignore

Reviewer: ```suggestion
### 2. Criação do arquivo .dockerignore
```

<details>
    <summary>

>       Clique aqui e inclua esse conteúdos no <code>.dockerignore</code>:

Reviewer: ```suggestion
      Clique aqui e inclua esse conteúdo no arquivo <code>.dockerignore</code>:
```

    </summary>
<div class="content">

```
.env
Dockerfile
.dockerignore
node_modules
npm-debug.log
README.md
.next
.git
```

</div>

</details>

> ### 3. Crie o Dockerfile

Reviewer: ```suggestion
### 3. Criação do arquivo Dockerfile
```

> > Como não estamos puxando as variáveis de ambiente do servidor para nosso contêiner, a [validação do esquema do ambiente](/en/usage/env-variables) falhará. Para evitar isso, temos que adicionar a flag `SKIP_ENV_VALIDATION=1` ao comando de compilação para que os schemas env não sejam validados no momento da compilação.

Reviewer: ```suggestion
> Como não estamos puxando as variáveis de ambiente do servidor para o nosso contêiner, a [validação do esquema do ambiente](/en/usage/env-variables) falhará. Para evitar isso, temos que adicionar a flag `SKIP_ENV_VALIDATION=1` ao comando de build para que os esquemas env não sejam validados no momento de build.
```

I changed in the suggestion `compilação` to `build`, because I thought in other places even in Portuguese in this PR, the word "build" is kept as is.
But I am not sure which would be better.

<details>
    <summary>

>       Clique aqui e inclua esse conteúdos no <code>.dockerignore</code>:

Reviewer: ```suggestion
      Clique aqui e inclua esse conteúdo no arquivo <code>Dockerfile</code>:
```

    </summary>
<div class="content">

```docker
##### DEPENDÊNCIAS

FROM --platform=linux/amd64 node:16-alpine AS deps
RUN apk add --no-cache libc6-compat openssl
WORKDIR /app

> # Instale o Prisma Client - remova se não estiver usando o Prisma

Reviewer 2: ```suggestion
# Instalação do Prisma Client - remova se não estiver usando o Prisma
```

COPY prisma ./

> # Instale dependências com base no gerenciador de pacotes preferido

Reviewer 2: ```suggestion
# Instalação de dependências com base no package manager padrão
```

 else echo "Lockfile not found." && exit 1; \
 fi

##### BUILDER

FROM --platform=linux/amd64 node:16-alpine AS builder
ARG DATABASE_URL
ARG NEXT_PUBLIC_CLIENTVAR
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

# ENV NEXT_TELEMETRY_DISABLED 1

RUN \
 if [ -f yarn.lock ]; then SKIP_ENV_VALIDATION=1 yarn build; \
 elif [ -f package-lock.json ]; then SKIP_ENV_VALIDATION=1 npm run build; \
 elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && SKIP_ENV_VALIDATION=1 pnpm run build; \
 else echo "Lockfile not found." && exit 1; \
 fi

##### RUNNER

FROM --platform=linux/amd64 node:16-alpine AS runner
WORKDIR /app

ENV NODE_ENV production

# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/next.config.mjs ./
COPY --from=builder /app/public ./public
COPY --from=builder /app/package.json ./package.json

COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs
EXPOSE 3000
ENV PORT 3000

CMD ["node", "server.js"]

```

> > **_Notas_**

Reviewer: ```suggestion
> **_Observações_**
```

>
> - _A emulação de `--platform=linux/amd64` pode não ser necessária após a mudança para o Node 18._
> - _Consulte [`node:alpine`](https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine) para entender por que `libc6-compat` pode ser necessário._

> > - _Next.js coleta [dados anônimos de telemetria sobre uso geral](https://nextjs.org/telemetry). Remova o comentário da primeira instância de `ENV NEXT_TELEMETRY_DISABLED 1` para desabilitar a telemetria durante a compilação. Remova o comentário da segunda instância para desabilitar a telemetria durante o tempo de execução._

Reviewer: ```suggestion
> - _Next.js coleta [dados anônimos de telemetria sobre uso geral](https://nextjs.org/telemetry). Remova o primeiro comentário de `ENV NEXT_TELEMETRY_DISABLED 1` para desabilitar a telemetria durante o build. Remova o segundo comentário para desabilitar a telemetria durante o tempo de execução._
```

</div>
</details>

> ## Crie e execute a imagem localmente

Reviewer: ```suggestion
## Execute build e run da imagem localmente
```
I opted for suggesting to keep the words "build" and "run" because they are the same as the commands.

Author: Fair enough

> Crie e execute esta imagem localmente com os seguintes comandos:

Reviewer: ```suggestion
Execute `build` e `run` desta imagem localmente com os seguintes comandos:
```

```bash
docker build -t ct3a-docker --build-arg NEXT_PUBLIC_CLIENTVAR=clientvar .
docker run -p 3000:3000 -e DATABASE_URL="database_url_goes_here" ct3a-docker
```

Abra [localhost:3000](http://localhost:3000/) para ver sua aplicação rodando.

## Docker Compose

Você também pode usar o Docker Compose para criar a imagem e executar o contêiner.

<details>
    <summary>

>       Siga as etapas 1 a 4 acima, clique aqui e inclua o conteúdo em <code>docker-compose.yml</code>:

Reviewer: ```suggestion
      Siga as etapas 1 a 4 acima, clique aqui e inclua o conteúdo no arquivo <code>docker-compose.yml</code>:
```

    </summary>
<div class="content">

```yaml
version: "3.9"
services:
  app:
    platform: "linux/amd64"
    build:
      context: .
      dockerfile: Dockerfile
      args:
        NEXT_PUBLIC_CLIENTVAR: "clientvar"
    working_dir: /app
    ports:
      - "3000:3000"
    image: t3-app
    environment:
      - DATABASE_URL=database_url_goes_here
```

Execute isso usando o comando `docker compose up`:

```bash
docker compose up
```

Abra [localhost:3000](http://localhost:3000/) para ver sua aplicação rodando.

</div>
</details>

## Deploy na Railway

> Você pode usar uma PaaS como [Railway's](https://railway.app) e seus [deploys de Dockerfile](https://docs.railway.app/deploy/dockerfiles) automáticos para fazer deploy das suas aplicações. Se você tiver o [CLI da Railway instalada](https://docs.railway.app/develop/cli#install), poderá implantar sua aplicação com os seguintes comandos:

Reviewer: ```suggestion
Você pode usar [deploys automáticos de Dockerfile](https://docs.railway.app/deploy/dockerfiles) em uma PaaS como a [Railway](https://railway.app) para fazer deploy das suas aplicações. Se você tiver o [CLI da Railway instalado](https://docs.railway.app/develop/cli#install), poderá fazer deploy da sua aplicação com os seguintes comandos:
```

```bash
railway login
railway init
railway link
railway up
railway open
```

> Vá para "Variáveis" e inclua seu `DATABASE_URL`. Em seguida, vá para "Configurações" e selecione "Gerar domínio". Para ver um exemplo em execução no Railway, visite [ct3a-docker.up.railway.app](https://ct3a-docker.up.railway.app/).

Reviewer: ```suggestion
Vá para "Variables" e inclua seu `DATABASE_URL`. Em seguida, vá para "Settings" e selecione "Generate Domain". Para ver um exemplo em execução no Railway, visite [ct3a-docker.up.railway.app](https://ct3a-docker.up.railway.app/).
```

I changed back the words to English because the user interface of Railway is in English.
